src/main/java/picocli/CommandTokenizer.java

```diff
@@ -0,0 +1,371 @@
+package picocli;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Scanner;
+import picocli.CommandLine;
+import picocli.CommandLine.Option;
+import picocli.CommandLine.Parameters;
+
+public class CommandTokenizer {
+
+		private String[] tokens;
+		private int position = 0;
+		// Characters that cimment out until a newline
+		private List<Character> commentChars = new ArrayList<Character>(Arrays.asList('#'));
+		// Characters that escape other special characters
+		private List<Character> escapeChars = new ArrayList<Character>(Arrays.asList('\\'));
+		// Characters to be escaped automatically
+		private List<Character> escapedChars = new ArrayList<Character>();
+		// Characters that begin/end quotes
+		private List<Character> quoteChars = new ArrayList<Character>(Arrays.asList('"', '\''));
+		// Characters to be treated as whitespace
+		private List<Character> whitespaceChars = new ArrayList<Character>(Arrays.asList(' '));
+		// Characters to use to represent a new line
+		private List<String> eolPatterns = new ArrayList<String>(Arrays.asList("\r\n", "\r", "\n"));
+		// Remove blanks at start and end of a string
+		private boolean trimBlanks = true;
+
+		public CommandTokenizer(String cmd) {
+			this.tokens = parse(cmd);
+		}
+
+		public CommandTokenizer(Scanner input) {
+			this.tokens = parse(input);
+		}
+
+		private String[] parse(String cmd) {
```

Comment:
Can we add unit tests for this method and the other public `parse` method? We can make this method package-private to allow unit tests to call it.

Reply:
Sure. I've never done unit testing before, but for something like this, I think it should be simple. 
I might actually make them `public`, so that people can reuse the object. In that case, I would also add a no-arg constructor.

Reply:
If you’ve never created unit tests before it will be a revelation. :-) Unit tests are a great way to shake out bugs and create confidence that future changes didn’t break anything. They are crucial to keeping software “soft”. 

Please see the picocli tests for examples.